neptune/observation.go

```diff
@@ -52,37 +51,35 @@ func buildObservationsQuery(instanceID string, f *observation.DimensionFilters)
 	}
 
 	var q string
-	var selectOpts []string
+	additionalDimensions := 0
+	var additionalDimensionOptions []string
 
 	for i, dim := range f.Dimensions {
 		if len(dim.Options) == 0 {
 			continue
 		}
 
-		for j, opt := range dim.Options {
-			dim.Options[j] = fmt.Sprintf("'%s'", opt)
+		optionIdPrefix := fmt.Sprintf("_%s_%s_", instanceID, dim.Name)
+		var optionIdList []string
+
+		for _, opt := range dim.Options {
+			optionId := optionIdPrefix + opt
+			optionIdList = append(optionIdList, fmt.Sprintf("'%s'", optionId))
 		}
 
-		// the first dimension filtered first, outside of the match statement
+		// the first dimension filtered independently of the rest to reduce the set of observations to filter, improving performance
 		if i == 0 {
-			q = fmt.Sprintf(query.GetFirstDimensionPart, instanceID, dim.Name, strings.Join(dim.Options, ","))
+			q = fmt.Sprintf(query.GetFirstDimensionPart, strings.Join(optionIdList, ","))
 			continue
 		}
 
-		// only add the match statement if there is a second dimension to match on
-		if i == 1 {
-			q += ".match("
-		}
-
-		selectOpts = append(selectOpts, fmt.Sprintf(query.GetObservationDimensionPart, instanceID, dim.Name, strings.Join(dim.Options, ",")))
-
+		additionalDimensions++
+		additionalDimensionOptions = append(additionalDimensionOptions, optionIdList...)
 	}
 
-	// only close the match statement if there are additional dimensions that have been matched
-	if len(f.Dimensions) > 1 {
-		//comma separate dimension option selections and close match statement
-		q += strings.Join(selectOpts, ",")
-		q += ")"
+	// only filter on additional dimensions if they are defined.
+	if additionalDimensions > 0 {
```

Comment:
is there actually any need to keep an `additionaDimensions` count? Couldnt we just check the len(addtionalDimensionOptions) > 0 ?

Reply:
or even just if len(f.Dimensions) >1 - which is likely a smaller array to check than the dimensionOptions itself

Reply:
`additionalDimensions` is mainly there to support the (magic) part of the query `.count(local).is_(%d)`. This is the count of dimensions being filtered on, minus the first one which is separate. I used it in the if statement as it seemed cleaner than the other options (it originally was `len(addtionalDimensionOptions) > 0`). 

Most of the time the number of `additionalDimensions` will be `len(f.Dimensions) - 1`, though I also wanted to cater for cases where not all dimensions are filtered.